src/lesson_4/FractionalWrapper.sol

```diff
@@ -0,0 +1,176 @@
+    constructor(IERC20 token, uint256 _exchangeRate, string memory name, string memory symbol) ERC20(name, symbol, 18) {
+        underlying = token;
+        exchangeRate = _exchangeRate;
+    }
+
+    ///@notice The address of the underlying token used for the vault for accounting, depositing, and withdrawing
+    ///@dev Will always be an ERC20 contract
+    ///@return assetTokenAddress address of the underlying token
+    function asset() public view returns (address assetTokenAddress) {
+        return address(underlying);
+    }
+
+    ///@notice Total amount of the underlying asset that is “managed” by the vault
+    ///@return totalManagedAssets balance of the underlying token within the vault
+    function totalAssets() public view returns (uint256 totalManagedAssets) {
+        return underlying.balanceOf(address(this));
+    }
+
+    ///@notice Calculates the amount of the asset (wrapped) token the user can get for their amount of the underlying
+    ///@param assets amount of the underlying token EX. DAI
+    ///@return shares amount of the asset (wrapped) token EX. fyDAI
+    function convertToShares(uint256 assets) public view returns (uint256 shares) {
+        return (assets * exchangeRate) / 1e27;
+    }
+
+    ///@notice Calculates the amount of the underlying token the user can get for their shares
+    ///@param shares amount of the asset (wrapped) token EX. fyDAI
+    ///@return assets amount of the underlying token EX. DAI
+    function convertToAssets(uint256 shares) public view returns (uint256 assets) {
+        return (shares * 1e27) / exchangeRate;
+    }
+
+    ///@notice Gives the maximum amount of assets that can be deposited by the receiver
+    ///@return maxAssets the maximum amount of assets available for deposit
+    function maxDeposit(address) public pure returns (uint256 maxAssets) {
+        return type(uint256).max;
+    }
+
+    ///@notice Gives the number of shares minted from a specified number of assets
+    ///@param assets the number of assets
+    ///@return shares the number of shares
+    function previewDeposit(uint256 assets) public view returns (uint256 shares) {
+        return convertToShares(assets);
+    }
+
+    ///@notice Mints shares for the receiver based on their assets
+    ///@param assets amount of the underlying asset
+    ///@param receiver the receiving address 
+    ///@return shares amount of the vault asset
+    function deposit(uint256 assets, address receiver) public returns (uint256 shares) {
+        shares = convertToShares(assets);
+        _mint(receiver, shares);
+        underlying.safeTransferFrom(msg.sender, address(this), assets);
+        emit Deposit(msg.sender, receiver, assets, shares);
+    }
+
+    ///@notice Gives the maximum amount of shares that can be minted by the receiver
+    ///@return maxShares the maximum amount of shares available for mint
+    function maxMint(address) public pure returns (uint256 maxShares) {
+        return type(uint256).max;
```

Comment:
Not true, `_totalSupply` can't exceed `type(uint256).max`

Reply:
Looking again at the EIP, would it simply be `return type(uint256).max - 1;`?

Reply:
I see.